### ISSUE TITLE
Manually specify table name

### ISSUE THREAD

**Issue by DeltaByte** [dynamoosejs/dynamoose#691]
### Summary:

Would it be a terrible idea to add a separate config for the table name inside of the schema definition? My specific use case is that the name of my Model does not always match the table name, especially since some of the tables that I use are for timeseries data, where I have a new table each month.

Ideally, the key would also accept a function that gets called at runtime for dynamic table names.

#### Schema
```
new dynamoose.Schema(
  {
    // attributes
  },
  {
    tableName: `prod_readings_NOV2019`,
    tableName: () => `${stage}_readings_${date}`
  }
)
```

#### Model
```
dynamoose.model('Reading', schema, options)
```

### Current output and behavior:

The table name is directly tied to the model name, and cannot change once compiled.

### Expected output and behavior:

Table name is a separate value to model name, though it still falls back to the model name as default behavior.

### Type (select 1):
- [ ] Bug report
- [x] Feature suggestion
- [ ] Question
- [ ] Other suggestion
- [ ] Something not listed here

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x] I have tested the code provided and am confident it doesn't work as intended
- [x] I have ensured that all of my plugins that I'm using with Dynamoose are listed above
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
@DeltaByte This will likely never be supported in Dynamoose. I don't think Mongoose supports this, and it seems unnecessary and not really taking into account the purpose of models and schemas.

Here would be my suggestion for how I would implement this.

Basically you can have the model be created dynamically as needed with a custom name. So something like the following:

**ReadingsModel.js**

```js
const dynamoose = require("dynamoose");
const schema = new dynamoose.Schema(
  {
    // attributes
  }
);

return (stage, date) => dynamoose.model(`${stage}_readings_${date}`, schema, options);
```

Then you can access it by doing the following:

**main.js**

```js
const Model = require("./ReadingsModel.js");
const model = Model("prod", "NOV2019");

model.get("item1", (item) => {
    console.log(item);
})
```

So basically you can create the models dynamically as needed by passing in the stage and date that you want.

I forget how Dynamoose handles duplicate model creation (ex. if you call `Model` with the exact same parameters multiple times), you might wanna test that and add a layer of caching to store models created into an object or something and check to see if they exist in that object and if so return that instead of calling `dynamoose.model`.

The biggest pitfalls in this design are the fact that since models are being created on the fly, you need to be careful around table creation and waitForActive settings on Dynamoose. Really think about how you want to verify the table is created and active here for tables that might not exist.

Hopefully that answers your question!

**Comment by DeltaByte** [comment]
@fishcharlie https://mongoosejs.com/docs/guide.html#collection

**Comment by dolsem** [comment]
@DeltaByte unless I missed the point @fishcharlie's approach works just fine to solve your problem. That being said, what's the motivation behind separating chunks of time series data into separate tables? I understand why you would want that with a traditional SQL database, but when it comes to DynamoDB I believe it is standard practice to keep it in a single table due to DynamoDB's properties. Instead separating data into multiple tables you might wanna separate into different shards instead.

**Comment by fishcharlie** [comment]
@DeltaByte I don’t see any reason for that property. It feels like it just adds a layer of truly unnecessary complexity. Although Mongoose parity is important, I’d need to understand more the reasoning behind their decision here.

Likely won’t happen for a while, but will likely reconsider at some point based on Mongoose’s support.

Everything @dolsem said applies too. Not sure why this additional layer of complexity is needed to achieve what you want.

**Comment by DeltaByte** [comment]
@dolsem While it isn't an issue now AWS has changed how they re-assign the capacity per partition, but historically the idea was to keep costs down by only keeping the frequently accessed data in a table with a large capacity assignment. The older, less accessed data could then live in tables with capacity that was decreased over time. (in an ideal world we would just refactor where the data is inserted from or delete it entirely, but that's not possible, unfortunately)

@fishcharlie You're correct in saying that the majority of people won't need that functionality, and if it doesn't get implemented then it's hardly a big issue. My reasoning for wanting it is mostly due to using babel with ES6 import/export syntax causing strange race-condition issues (especially with mocking config), and some potential issues with memory usage in situations when the application is long-lived and  changes tables frequently (meaning lots of cached models)

**Comment by fishcharlie** [comment]
@DeltaByte So for memory concerns, I’m still not sure how adding this property would help. Dynamoose will still have to do the work of creating the table and ensuring its active. There is just more overhead for new tables in DynamoDB vs new collections in MongoDB.

**Comment by Chrisissorry** [comment]
@fishcharlie I came here because I use serverless framework. There is a way to specify the table name in the serverless config. There, you can access all env variables and use that for the naming, e.g. stage and project name. Now, the whole config for the dynamoDB table can be done in serverless, but the table name cannot be passed to dynamoose. Doing that via the model name just feels awkward. Especially because `aws-sdk` offers that ouf ot the box.

**Comment by fishcharlie** [comment]
@Chrisissorry 

> Doing that via the model name just feels awkward

Kinda. But remember in v2 a model is meant to represent a Table.

In v3, this is changed to where you can create a `Table` instance. By default to ensure backwards compatibility and keep things simple, a Model will automatically create a Table when you first run a request on it (#1453). However, you can manually attach a Model to a Table, which makes this more clear.

---

Beyond that I'm not sure I completely understand what else you are asking for here.
